Speed up computation of convergent fractions and unify it

HannesWell opened the pull request:
This PR has the goal to speed up and to unify the computation of fraction convergents by avoiding the creation of unnecessary (Big-)Fraction objects. Furthermore this change allows to specify arbitrarily precise convergence-tests.
Instead of a Stream of converging `(Big-)Fractions` only the last object is returned. This makes it more elegant in some scenarios but less elegant in others.
The method returns a `(Big-)Fraction` in any case, either because the `maxIteration` limit was exceeded or because the convergence-test passed, which is necessary for those cases that are iteration bounded.

Parts of the code (especially documentation) are not yet final respectively have to be adjusted, but I would like to agree on the technical details first.

@maisonobe I'm looking forward to your review.
/CC @axkr

maisonobe commented on hipparchus-core/src/main/java/org/hipparchus/fraction/BigFraction.java:
```diff
@@ -338,23 +339,19 @@ public BigFraction(final long num, final long den) {
      * @return stream of {@link BigFraction} convergents approximating  {@code value}
      * @since 2.1
      */
-    public static Stream<BigFraction> convergents(final double value, final int maxConvergents) {
```
Could we just keep both the existing `convergents` method which provides all intermediates in addition to the new `convergent` method which avoids building these intermediate terms? They seem to answer different use cases to me.
Of course, the constructors above that only need the last convergent should use the new method as you propose.

I see in the tests that it is possible to use `convergent` to keep the intermediate by adding them as a side effect of checking convergence, but it seems awkward to me, so I would prefer to have both methods.

What do you think?

HannesWell replied:
> 
> What do you think?

Fully agree. I didn't liked the side-effects as well.

HannesWell commented:
I reworked this PR to keep the stream of convergent and to add another method that only returns the last convergent methods available. To avoid code duplication I added the package private class ConvergenceStep that is used by both methods.

This change is also less strict when it comes to detection of overflows for BigFractions and does not generally throw an overflow exception if a value in the fraction is greater than Integer.MAX_VALUE. To still detect overflows I use the `strictAdd/Multiply` methods instead.

I also  created a simple poor-mans benchmark as Test-Case in `BigFractionTest` to get an rough estimate about the runtime improvements of the new implementation:

```
   @Test
    public void testApproximationRuntime() {

        System.out.println("///////////////////////");
        System.out.println("Random test start");
        System.out.println("///////////////////////");

        int notExact = 0;

        Random rnd = new Random();
        int maxExp = 10;
        double[] randomValues = IntStream.range(0, 10_000_000)
                .mapToDouble(i -> rnd.nextDouble() * Math.pow(10, rnd.nextInt(2 * maxExp) - maxExp)).toArray();
        long start = System.currentTimeMillis();
        for (double val : randomValues) {
            notExact += checkValue(val);
        }
        System.out.println("Done. Not exakt: " + notExact + " Took " + (System.currentTimeMillis() - start));
    }

    private static int checkValue(double val) {
        try {
            BigFraction fraction = new BigFraction(val, 0.5E-4, 20);
            return isExact(val, fraction);
        } catch (Exception e) {
            e.printStackTrace();
            return 1;
        }
    }

    private static int isExact(double val, BigFraction approximation) {
        double approximationValue = approximation.doubleValue();
        if (Double.doubleToLongBits(approximationValue) != Double.doubleToLongBits(val)) {
            return 1;
        }
        return 0;
    }
```
For multiple runs the print-out displays for the old implementation a runtime about 3300~3600ms and for the new implementation ~800ms. If you want a more reliable benchmark I can set up an JMH case.

HannesWell commented on hipparchus-core/src/main/java/org/hipparchus/fraction/ConvergentsIterator.java:
```diff
@@ -21,112 +21,97 @@
-        stop = Precision.equals(convergent, value, 1);
-        if ((p2 > OVERFLOW || q2 > OVERFLOW) && !stop) {
-            throw new MathIllegalStateException(LocalizedCoreFormats.FRACTION_CONVERSION_OVERFLOW, value, p2, q2);
+    static Pair<ConvergenceStep, Boolean> convergent(double value, int maxIterations,
```
Do you think this method should just return a ConvergenceStep without the flag that indicated if the value has converged?
I added it with the intention that users could only want a value if it has converged but at least the direct usage in (Big-)Fraction most of the time requires just the last value regardless if hat has passed the convergence test.

If just the step is returned it should be stated in the doc (I still have to write) that the returned fraction is not guaranteed to fulfill the convergence-test and the convergence-test is more a criteria to abort the iteration prematurely.

maisonobe replied:
No, your design is fine to me.

So I just wait for the doc and will merge the PR.

Thanks again for your contribution!

HannesWell replied:
> No, your design is fine to me.
> 
> So I just wait for the doc and will merge the PR.

Done.
> 
> Thanks again for your contribution!

You are very welcome and thanks for Hipparchus!

maisonobe commented on hipparchus-core/src/main/java/org/hipparchus/fraction/BigFraction.java:
```diff
@@ -267,16 +267,19 @@ public BigFraction(final double value, final double epsilon,
      *             if the continued fraction failed to converge.
      */
-    public BigFraction(final double value, final int maxDenominator)
+    public BigFraction(final double value, final int maxDenominator) // TODO: long maxDenominator?
```
I also think a long would be better here

HannesWell replied:
I adjusted it accordingly.

HannesWell commented:
First of all, sorry for the delay.
I have added java-doc to all new methods and FunctionalInterfaces and fixed a minor inconsistency in the returned Pair's value (which was always false if the element at position `maxConvergents` was examined even if it actually satisfies the convergents-test).

Please let me know if you have any suggestions for improvements or the doc does not match the usual style of Hipparchus.

maisonobe commented:
I have merged the request, and added one change. I had to use `Spliterators.spliteratorUnknownSize` again because
`Stream.iterate` with a predicate to stop iterations has been introduced in Java 9 and Hipparchus is still stuck to Java 8.
I know this is a really old version, and would like to jump to something more modern (probably Java 17), but cannot do that now.
I then added a few more javadoc and moved around some methods to fix issues detected by our [SonarQube instance](https://sonar.orekit.org/dashboard?id=org.hipparchus%3Ahipparchus-aggregator).

Thanks a lot for this contribution!

HannesWell commented:
Thank you for you review and the subsequent clean-up!

I have to admit that I did not check the Java-Version Hipparchus is relaying on. I'm using Java-11 a.t.m. but hope to be able to move to 17 soon and I just took what my IDE offered me. But you are not the last library using Java-8. However moving forward is definitely good. :)
If you would have set up a build workflow that is triggered on PR pushes similar to your main workflow that reacts to pushes on master (just like this https://github.com/axkr/symja_android_library/blob/master/.github/workflows/maven-build-pr.yml) I would have noticed this earlier and would have fixed it.


> I then added a few more javadoc and moved around some methods to fix issues detected by our [SonarQube instance](https://sonar.orekit.org/dashboard?id=org.hipparchus%3Ahipparchus-aggregator).

Unfortunately the SonarLint Eclipse Plug-in I have installed did not marked this issues locally.
But having a SonarQube instance set up is great! Are they free of charge for OSS projects?

> Thanks a lot for this contribution!

You are very welcome :)

maisonobe commented:
This kind of fix is not a problem, I was very happy with what you already did!
This SonarQube instance is maintained by the company I work for. It is a part of the continuous integration system we have set up around the three open-source projects Rugged, Orekit and Hipparchus in which my company is deeply involved. It is not intended for other projects, we are not a hosting company but an engineering company working among others in the space field.

thanks again!

HannesWell commented:
Glad to hear that. :)

Understand it. In the company I work for, we also has a SonarQube-Instance, but it's just internal. However I think having a SonarQube would also be good for OSS projects and therefore keep my eyes open for opportunities. I found the [Sonarcloud-Scan Action](https://github.com/marketplace/actions/sonarcloud-scan) for GH which states to provide a free of charge service for public OSS projects. I first assumed you are using this (or a similar service) and I hoped to gather experience.
Nevertheless its great your company supports these projects. 👍🏽 
And thanks for your elaboration!
